signer/plugin.go

```diff
@@ -0,0 +1,259 @@
+	"errors"
+	"fmt"
+
+	"github.com/notaryproject/notation-core-go/signature"
+	"github.com/notaryproject/notation-go"
+	"github.com/notaryproject/notation-go/internal/envelope"
+	"github.com/notaryproject/notation-go/plugin"
+	"github.com/notaryproject/notation-go/plugin/proto"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+)
+
+// pluginSigner signs artifacts and generates signatures.
+// It implements notation.Signer
+type pluginSigner struct {
+	plugin       plugin.SignPlugin
+	keyID        string
+	pluginConfig map[string]string
+	signer       signature.Signer
+}
+
+// NewSignerPlugin creates a notation.Signer that signs artifacts and generates signatures
+// by delegating the one or more operations to the named plugin, as defined in
+// https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md#signing-interfaces.
+func NewFromPlugin(plugin plugin.Plugin, keyID string, pluginConfig map[string]string) (notation.Signer, error) {
+	if plugin == nil {
+		return nil, errors.New("nil plugin")
+	}
+	if keyID == "" {
+		return nil, errors.New("nil signing keyID")
+	}
+	// remoteSigner implements signature.Signer
+	remoteSigner := &remoteSigner{
+		plugin:       plugin,
+		pluginConfig: pluginConfig,
+		keyID:        keyID,
+	}
+	return &pluginSigner{
+		plugin:       plugin,
+		keyID:        keyID,
+		pluginConfig: pluginConfig,
+		signer:       remoteSigner,
+	}, nil
+}
+
+// Sign signs the artifact described by its descriptor and returns the marshalled envelope.
+func (s *pluginSigner) Sign(ctx context.Context, desc ocispec.Descriptor, opts notation.SignOptions) ([]byte, *signature.SignerInfo, error) {
+	req := &proto.GetMetadataRequest{
+		PluginConfig: s.pluginConfig,
+	}
+	metadata, err := s.plugin.GetMetadata(ctx, req)
+	if err != nil {
+		return nil, nil, err
+	}
+	if !metadata.SupportsContract(proto.ContractVersion) {
+		return nil, nil, fmt.Errorf(
+			"contract version %q is not in the list of the plugin supported versions %v",
+			proto.ContractVersion, metadata.SupportedContractVersions,
+		)
+	}
+	if metadata.HasCapability(proto.CapabilitySignatureGenerator) {
+		return s.generateSignature(ctx, desc, opts)
+	} else if metadata.HasCapability(proto.CapabilityEnvelopeGenerator) {
+		return s.generateSignatureEnvelope(ctx, desc, opts)
+	}
+	return nil, nil, fmt.Errorf("plugin does not have signing capabilities")
+}
+
+func (s *pluginSigner) generateSignature(ctx context.Context, desc ocispec.Descriptor, opts notation.SignOptions) ([]byte, *signature.SignerInfo, error) {
+	config := s.mergeConfig(opts.PluginConfig)
+	// Get key info.
+	key, err := s.describeKey(ctx, config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Check keyID is honored.
+	if s.keyID != key.KeyID {
+		return nil, nil, fmt.Errorf("keyID in describeKey response %q does not match request %q", key.KeyID, s.keyID)
+	}
+	ks, err := proto.DecodeKeySpec(proto.KeySpec(key.KeySpec))
```

Comment:
We can update `proto.DescribeKeyResponse`.KeySpec type and remove this type conversion.